drivers/retained_mem/retained_mem_nrf_gpregret.c

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2023, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_nrf_gpregret
+
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/retained_mem.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(retained_mem_nrf_gpregret, CONFIG_RETAINED_MEM_LOG_LEVEL);
+
+#ifdef CONFIG_MULTITHREADING
+struct nrf_gpregret_data {
+	struct k_mutex lock;
+};
+#endif
+
+struct nrf_gpregret_config {
+	uint8_t *addr;
+	size_t size;
+};
+
+static int32_t nrf_gpregret_init(const struct device *dev)
+{
+#ifdef CONFIG_MULTITHREADING
+	struct nrf_gpregret_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
+#endif
+
+	return 0;
+}
+
+static ssize_t nrf_gpregret_size(const struct device *dev)
+{
+	const struct nrf_gpregret_config *config = dev->config;
+
+	return (ssize_t)config->size;
+}
+
+static int32_t nrf_gpregret_read(const struct device *dev, size_t offset, uint8_t *buffer,
+				 size_t size)
+{
+	const struct nrf_gpregret_config *config = dev->config;
+
+#ifdef CONFIG_MULTITHREADING
+	struct nrf_gpregret_data *data = dev->data;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+#endif
+
+	memcpy(buffer, (void *)(config->addr + offset), size);
```

Comment:
nonblocking: you can drop the cast now that you've changed `addr` to `uint8_t*`

Reply:
Can't reply to the above comment for some reason but `int` is not 64-bit on 64-bit builds, `long` is a 64-bit value:
```
printk("int = %d\n", sizeof(int));
```
=
```
Hello World! native_posix_64
int = 4
```